libusb/hid.c

```diff
@@ -276,6 +285,81 @@ static int get_usage(uint8_t *report_descriptor, size_t size,
 	return -1; /* failure */
 }
 
+/* Retrieves the largest report size (in bytes) from the passed in report descriptor.
+   The return value is the size on success and -1 on failure. */
+static size_t get_max_report_size(uint8_t * report_descriptor, int desc_size, enum report_descr_type report_type)
+{
+	int i = 0;
+	int size_code;
+	int data_len, key_size;
+
+	int64_t report_size = -1, report_count = -1;
+	size_t cur_size = 0;
+	size_t max_size = 0;
+
+	while (i < desc_size) {
+		int key = report_descriptor[i];
+		int key_cmd = key & 0xfc;
+
+		if ((key & 0xf0) == 0xf0) {
+			/* This is a Long Item. The next byte contains the
+			   length of the data section (value) for this key.
+			   See the HID specification, version 1.11, section
+			   6.2.2.3, titled "Long Items." */
+			if (i+1 < desc_size)
+				data_len = report_descriptor[i+1];
+			else
+				data_len = 0; /* malformed report */
+			key_size = 3;
+		} else {
+			/* This is a Short Item. The bottom two bits of the
+			   key contain the size code for the data section
+			   (value) for this key.  Refer to the HID
+			   specification, version 1.11, section 6.2.2.2,
+			   titled "Short Items." */
+			size_code = key & 0x3;
+			data_len = (size_code < 3) ? size_code : 4;
+			key_size = 1;
+		}
+
+		if (key_cmd == 0x94) { /* Report Count */
+			report_count = get_bytes(report_descriptor, desc_size, data_len, i);
+		}
+		if (key_cmd == 0x74) { /* Report Size */
+			report_size = get_bytes(report_descriptor, desc_size, data_len, i);
+		}
+		if (key_cmd == report_type) { /* Input / Output / Feature */
+			if (report_count < 0 || report_size < 0) {
+				/* We are missing size or count. That isn't good. */
+				return 0;
```

Comment:
```suggestion
			if (report_count < 0 || report_size < 0) {
				/* We are missing size or count. That isn't good. */
				return -1;
```
This would mean a corrupt ReportDescriptor

Reply:
I wasn't consistent with whether I was using `size_t` or `ssize_t`. In my last commit, I settled on `size_t`, but perhaps I should change it back to `ssize_t` so that we can more clearly differentiate between errors and a zero value (if there are no feature reports, it will return 0, which is not an error).

Reply:
Makes sense